### PR TITLE
Add a new `hostStats` query for statistics counts

### PR DESCRIPTION
**Pull request by PaulWay**
The Insights Dashboard makes three requests to the Inventory API to gather counts of how many hosts are current, stale and stale-warn.  To support this we provide the `hostStats` query, which aggregates these counts and takes a set of `hostFilter` filters.

The `hostStats/index.js` code is still in the early phases, as I'm unfamiliar with Elasticsearch and only moderately familiar with GraphQL and TypeScript.  I'd welcome any assistance with getting this working!

### CODE REVIEW

**Review by thearifismail — COMMENTED** [review]
LGTM but needs some tests

**Comment by thearifismail** [comment]
@PaulWay I hope you can reopen this PR.  All I did was comment on the PR to provide tests and submitted the review but now I see the PR as "Closed".  

Have I stepped on some hidden bug.  The same thing happened to me a few days ago.

**Comment by kruai** [comment]
> @PaulWay I hope you can reopen this PR. All I did was comment on the PR to provide tests and submitted the review but now I see the PR as "Closed".
> 
> Have I stepped on some hidden bug. The same thing happened to me a few days ago.

@thearifismail  This PR isn't closed; the entry above yours is saying that this PR was referenced by HBI PR 1264, which is now closed

File: src/resolvers/hostStats/index.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    const dataSelectionSet = _.find(selectionSet, s => s.name.value === 'data');
+    // const sourceList: string = dataSelectionSet.selectionSet.selections.map(
+    //     (o: any) => o.name.value
+    // ) /* .map(translateFilterName) ? */;
+
+    // const query: any = {
+    //     track_total_hits: true,
+    //     _source: sourceList
+    // };
```
Comment (kruai) [inline src/resolvers/hostStats/index.ts:23]: Commented code blocks should be removed

File: src/resolvers/hostStats/index.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                },
+                filter: {
+                    stale_warning_timestamp: {
+                        lt: now
+                    }
+                }
+            }
+        }
+    };
+    // query.query = buildFilterQuery(args.filter, org_id);
```
Comment (kruai) [inline src/resolvers/hostStats/index.ts:61]: Remove

**Review by kruai — CHANGES_REQUESTED** [review]
A couple of comments. I'm pulling this to do some local testing, but this should also have a couple of tests (probably in a file called `hostStats.integration.ts`). 
Just to confirm, do you have it working yet? If so, can you paste an example query so I can test it locally?

File: src/resolvers/hostStats/index.ts (hunk 1 above)
Comment (PaulWay) [reply src/resolvers/hostStats/index.ts:23]: Hi @kruai - appreciate that.  I was leaving the code commented while I worked out what I needed for the actual ElasticSearch query.  If I needed it back, then I hadn't lost it entirely :-)

What I really need is to know whether the ElasticSearch query actually works, and whether it gets the statistics the Dashboard needs - the total number of hosts, and the number of stale and stale warning hosts, meeting the given filter terms.  This should take exactly the same set of filters that the `hosts` query does.  Is the ElasticSearch query working?

Comment (kruai) [reply src/resolvers/hostStats/index.ts:23]: I haven't been able to get it working locally, no - I was hoping you had tried it out locally as you were coding it. I'll debug locally a bit more and see if it's apparent where the issue is occurring

Comment (kruai) [reply src/resolvers/hostStats/index.ts:23]: The most info I can get is `Cannot return null for non-nullable field Query.hostStats.`
It makes sense that the field is non-nullable since you want to get something back each time you query it, but I think something in the resolver is returning null.

**Comment by PaulWay** [comment]
Are there any other XJoin developers that can have a look at this?  This seems to be waiting a long time?

**Comment by fijshion** [comment]
yea I can take a look. I saw some open ended conversations so I thought it was being handled already. Is this still on your radar @kruai ?

**Comment by kruai** [comment]
> yea I can take a look. I saw some open ended conversations so I thought it was being handled already. Is this still on your radar @kruai ?

I replied in the review thread that the query is not working, and that work needed to be done in the resolver. I didn't realize you were expecting me to resolve that on my end @PaulWay - sorry for that miscommunication :sweat_smile: 

To answer your question, I haven't been working on this since that initial debug. I was waiting for another reply or code update before revisiting it

**Comment by PaulWay** [comment]
Hi @fijshion - yeah, it really needs your help :-)

We're trying to replace the three calls to Inventory that just fetch the count of the number of total, stale and stale-warning hosts.  So the expected response is something like:

```json
{
    "total_hosts": 365,
    "stale_hosts": 205,
    "warn_hosts": 47,
}
```

The idea is to maybe be able to expand with other counts - e.g. the count of Ansible and MS SQL hosts - to further reduce the number of requests being made to the Inventory API.

**Comment by PaulWay** [comment]
Hi @fijshion @kruai et al - I've tried but I cannot set up the elasticsearch container.  I can't find a copy of the `docker` command that runs on RHEL 8 CSB, `podman-compose` doesn't like bringing up just one container, and when I try just `podman-compose up`it gives me permission denied errors.  I've posted about this a number of times in emails and in Slack but received no response.

**Comment by PaulWay** [comment]
Can you maybe give more information about what your development and test environments are, so that I can try to duplicate them?

**Comment by fijshion** [comment]
@PaulWay I use docker-compose. Have you tried installing docker via this? https://docs.docker.com/engine/install/centos/

File: src/resolvers/hostStats/index.ts (hunk 2 above)
Comment (PaulWay) [reply src/resolvers/hostStats/index.ts]: Fixed.

File: src/resolvers/hostStats/index.ts (hunk 1 above)
Comment (PaulWay) [reply src/resolvers/hostStats/index.ts]: Turns out that the hostStats resolver wasn't in the list of resolvers.

**Review by strider — APPROVED** [review]
Tested new search locally. LGTM, but would prefer to get another pair of eyes on this one!

@kruai @fijshion @thearifismail

**Comment by PaulWay** [comment]
Hi @fijshion and @strider  - thanks for the reviews!  I've squash rebased it all down to one commit now.

Not sure what kind of tests you're wanting there @fijshion - are you wanting tests where we change the staleness of some hosts and see if the counts change?  Or something else?

**Comment by PaulWay** [comment]
BTW, if that rebase failed, and it looks like it's baked in some master commits that might confuse things, I do have a branch at the point in the commit log just before I rebase-squashed it.  So I can redo this if we need to :grinning:

**Comment by fijshion** [comment]
@PaulWay yea a test that validates the correct counts are returned for stale/warn when there is at least one of each

**Comment by PaulWay** [comment]
OK, I've modified my test data to have some different staleness dates - let's see how the tests go...

**Review by fijshion — APPROVED** [review]

**Comment by PaulWay** [comment]
Thanks @fijshion !
